Add Netlify deploy guide (#1733)
apollographql/apollo-server: docs/source/deployment/netlify.md

---

> title: Deploy a fullstack Apollo app with Netlify

Reviewer: ```suggestion
title: Deploying Apollo Server to Netlify
```

sidebar_title: Netlify
description: How to run your API using Netlify Functions on AWS Lambda
---

> [Netlify functions](https://www.netlify.com/docs/functions/), is a new way to deploy server-side code to AWS Lambda alongside your frontend, right from GitHub. In this guide, I'll show you why Netlify's functions is a great way to run your whole Apollo app.

Reviewer: ```suggestion
[Netlify Functions](https://www.netlify.com/docs/functions/) allow deploying server-side code, directly from GitHub, to AWS Lambda.  This guide demonstrates how to use Netlify Functions to deploy a GraphQL server and how a React application (built with `create-react-app`) can consume data from that server.
```

1. Use the same platform to deploy your frontend and API code.
2. Deploy previews of both your frontend and API on every pull request.

> 3. No need to worry about scaling containers due to a serverless architecture.

Reviewer: ```suggestion
3. No need to worry about scaling containers due to a "serverless" infrastructure.
```

We'll set up a "Hello World" API and frontend for an Apollo app on Netlify. Let's get started!

## Download and Run the Netlify Starter Kit

> Download and run the [official Netlify create-react-app and Lambda starter kit](https://github.com/netlify/create-react-app-lambda). This will set us up with a base for our frontend and API code.

Reviewer 2: Super awesome they already upgraded this to CRAv2

> Download and run the [official Netlify create-react-app and Lambda starter kit](https://github.com/netlify/create-react-app-lambda). This will set us up with a base for our frontend and API code.

Reviewer: ```suggestion
Download and run the [official Netlify `create-react-app` and Lambda starter kit](https://github.com/netlify/create-react-app-lambda). This will set us up with a base for our frontend and API code.
```

```bash

> git clone https://github.com/netlify/create-react-app-lambda.git

Reviewer 2: Could this perhaps be easier to digest using [`npm init`'s `initializer`](https://docs.npmjs.com/cli/init#examples)?

```bash
npm init react-app-lambda
```

Reviewer: Hmm, I'm not sure how this would currently work since [`create-create-app-lambda` isn't an npm package](http://npm.im/create-react-app-lambda), merely a Git repo.  Additionally, I don't believe that initializers work in any sort of automatic cloning way?  A script would have to be specified as its main `bin` which did the cloning — right?  For example, [`create-esm`'s main `bin`](https://github.com/standard-things/create-esm/blob/c2e1ff3e19cf2ec478bb93ed857d6b0e52cb6ff0/package.json#L9) is the `create-esm` script.  That script, [does a bunch of things](https://github.com/standard-things/create-esm/blob/c2e1ff3e19cf2ec478bb93ed857d6b0e52cb6ff0/src/create-esm.js) to make it work.  Unless I'm missing something?

Reviewer 2: Good point @abernix about `create-react-app-lambda` not being an npm package, not sure if it would work.

I assumed it would because the repo is based on CRAv2 and CRA recommends using the [`npm init react-app` pattern](https://github.com/facebook/create-react-app#npm)

I didn't realized it until now; that will require npm 6, which obviously isn't a minimum for these docs.

cd create-react-app-lambda
npm install
```

Let’s take a quick look at the file structure here:

> * **package.json:** our dependencies, shared between client and server code. You can also split them up into separate directories later, depending on your preference.

Reviewer: ```suggestion
* **package.json:** This application's dependencies, shared between client and server code. You can also split them up into separate directories later, depending on your preference.
```

> * **netlify.toml:** configuration for Netlify and the Netlify functions.

Reviewer: ```suggestion
* **netlify.toml:** The configuration for Netlify.
```

> * **src/:** code for our React app.

Reviewer: ```suggestion
* **src/:** The source code for the React frontend app.
```

> * **src/lambda/:** code that will be deployed to Lambda by Netlify.

Reviewer: ```suggestion
* **src/lambda/:** The server source code that will be deployed to Netlify Functions.
```

> In `package.json`, you’ll find scripts to run the frontend and lambda code:

Reviewer: ```suggestion
In `package.json`, you'll find scripts to run the frontend and lambda code:
```

```js
"scripts": {
  "start": "react-scripts start",
  "start:lambda": "netlify-lambda serve src/lambda",
  "build": "react-scripts build",
  "build:lambda": "netlify-lambda build src/lambda",
}
```

Run the code to see that everything is working. We have to open two terminal windows to run the frontend and functions at the same time:

```bash

> # In the first terminal

Reviewer: ```suggestion
# In the first terminal, run the local development server.
# This emulates the Netlify Functions runtime environment.
```

> npm run start:lambda

Reviewer: I don't understand what these commands are going to do.  Can we explain that a bit?  Am I actually going to start paying money with `npm run start:lambda`?  Does it actually deploy to Netlify at that point?

Can we provide some details about the difference between `start` and `build`?

> # In the second terminal

Reviewer: ```suggestion
# In the second terminal, start the React frontend application.
```

npm start
```

> If everything looks like it started correctly, let’s add a GraphQL API with Apollo Server!

Reviewer: ```suggestion
If everything looks like it started correctly, let's add a GraphQL API with Apollo Server!
```

### Setting up Apollo Server

> Netlify functions run on AWS Lambda, so we can use the `apollo-server-lambda` package to easily integrate Apollo Server for our API layer. Let’s install the packages we need for that:

Reviewer: ```suggestion
Netlify Functions run on AWS Lambda, so we can use the `apollo-server-lambda` package to easily integrate Apollo Server for our API layer. Let’s install the packages we need for that:
```

```bash
npm install --save apollo-server-lambda graphql
```

> Now, we can create a “Hello world” GraphQL API. Let’s put that in a new file in the lambda folder, at `src/lambda/graphql.js`:

Reviewer: ```suggestion
Now, we can create a "Hello world" GraphQL API. Let's put that in a new file in the `lambda` folder, at `src/lambda/graphql.js`:
```

```js
// src/lambda/graphql.js
const { ApolloServer, gql } = require("apollo-server-lambda");

const typeDefs = gql`
  type Query {
    hello: String
  }
`;

const resolvers = {
  Query: {

>     hello: (root, args, context) => {

Author: @abernix What do you mean by it's not always "root" ? Please can I get a verbose explanation or a link to an article that explains why?

>     hello: (root, args, context) => {

Reviewer: It's not `root` because revolvers are called in a cascading fashion.  This is only "`root`" it if it's at the very top of the graph entry point.

What it really points to is the resolver that called it — it's parent — and in the case of the first resolver being called, that parent is the `Query` type.

>     hello: (root, args, context) => {

Reviewer: ```suggestion
    hello: (parent, args, context) => {
```

      return "Hello, world!";
    }
  }
};

const server = new ApolloServer({
  typeDefs,
  resolvers
});

exports.handler = server.createHandler();
```

> Now, make sure you’ve run `npm run start:lambda`, and navigate to `localhost:9000/graphql` in your browser. You should see GraphQL Playground, where you can run queries against your API!

Reviewer: nit: `you've` instead of `you’ve` (fancy apostrophe?) Same for other places?

> Now, make sure you’ve run `npm run start:lambda`, and navigate to `localhost:9000/graphql` in your browser. You should see GraphQL Playground, where you can run queries against your API!

Reviewer: ```suggestion
Now, make sure you've run `npm run start:lambda`, and navigate to `localhost:9000/graphql` in your browser. You should see GraphQL Playground, where you can run queries against your API!
```

<div style="text-align:center">
![Local GraphQL Server](../images/graphql.png)
<br></br>
</div>

> If you can see GraphQL Playground and run a simple query, you’ve done everything properly! It’s pretty neat that getting an API up and running was just a few lines of code. Now let’s add Apollo to our frontend.

Reviewer: ```suggestion
If you can see GraphQL Playground and run a simple query, you've done everything properly. Now, let's add Apollo Client to the frontend.
```

> ### Adding Apollo Client to React

Reviewer: Hmm.  I haven't checked if we're doing this elsewhere in the server section, but I don't think demonstrating client details belongs in this guide.  By putting it here, it makes management of the documentation more difficult since those that are responsible for the Apollo Client changes won't be able to easily locate and fix these as client versions evolve.

> Open up `src/App.js`. Let’s convert this file to load a simple query from the API we just set up. First, let’s install Apollo Client and the React integration, and we’ll use the `apollo-boost` package to get started easily:

Reviewer: ```suggestion
Open up `src/App.js`. Let's convert this file to load a simple query from the API we just set up. First, let's install Apollo Client and the React integration, and we’ll use the `apollo-boost` package to get started easily:
```

```bash
npm install --save apollo-boost react-apollo
```

Now, add some code to initialize Apollo Client:

```js
// src/App.js
import ApolloClient from "apollo-boost";
const client = new ApolloClient({
  uri: "/.netlify/functions/graphql"
});
```

The way that the starter kit and Netlify are set up, any requests that start with `.netlify/functions` are automatically redirected to the Lambda functions in our project. So since we created our API with the filename `graphql.js`, we can call it with `.netlify/functions/graphql` from our frontend.

> Let’s replace the LambdaDemo component from the starter kit with a new one that uses Apollo to query our API:

Reviewer: ```suggestion
Let's replace the LambdaDemo component from the starter kit with a new one that uses Apollo to query our API:
```

```js
import { gql } from "apollo-boost";
import { ApolloProvider, Query } from "react-apollo";
// Replace the previous LambdaDemo with the code below:
const LambdaDemo = () => (
  <ApolloProvider client={client}>
    <Query
      query={gql`
        {
          hello
        }
      `}
    >
      {({ data }) =>
        <div>A greeting from the server: {data.hello}</div>}
    </Query>
  </ApolloProvider>
);
```

> If we’ve got everything set up correctly, we’ll see the greeting that we wrote in our resolver show up in the frontend:

Reviewer: ```suggestion
If we've got everything set up correctly, we'll see the greeting that we wrote in our resolver show up in the frontend:
```

<div style="text-align:center">
![React Frontend](../images/welcometoreact.png)
<br></br>
</div>

> Sweet, now we’ve connected our client to our API, and the only thing left is to deploy it! We can deploy both our frontend and API to Netlify in just one step.

Reviewer: I haven't read the above client code because of my comments above, but I can re-review if we decide to keep it.

### Deploying to Netlify

> The best way to deploy to Netlify is to set it up to deploy automatically from a GitHub repository. Therefore, create a new repository and push the code there.

Reviewer: I'm not sure we're in a position to say what's best for Netlify.  Can we word this differently?

> Netlify supports automatically deploying from a GitHub repository. Therefore, ...

> The best way to deploy to Netlify is to set it up to deploy automatically from a GitHub repository. Therefore, create a new repository and push the code there.

Reviewer: ```suggestion
Netlify can deploy automatically from a linked GitHub repository, so we'll use that technique for the rest of this tutorial which will allow each `git push` to automatically deploy. To enable this, first [create a new, empty repository on GitHub](https://help.github.com/articles/create-a-repo/).
```

To start the project, we cloned the Netlify starter kit. If we want the local repository to point to our new GitHub repo instead of the starter kit, we need to change the `origin`, like this:

```bash

> # Replace the URL with the one to your repository, example:

Reviewer: ```suggestion
# Replace the following URL with the appropriate URL obtained when
# creating the new GitHub repository in the previous step!
```

> git remote set-url origin git@github.com:stubailo/apollo-netlify-lambda-app.git

Reviewer: Hmm, this would work, but if I was using a boilerplate I'd probably just delete the `.git` directory.  That's maybe just me though.

> git remote set-url origin git@github.com:stubailo/apollo-netlify-lambda-app.git

Reviewer: ```suggestion
git remote set-url origin git@github.com:username/repository.git
```
